src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs

```diff
@@ -303,120 +372,411 @@ public void EnumerableStopsAtEndWhenEndIsFirstByteOfEmpty()
+        {
+            ReadOnlySequence<byte> buffer = ReadOnlySequence<byte>.Empty;
+            Assert.Throws<ArgumentOutOfRangeException>("offset", () => buffer.Slice(0, 1));
+            Assert.Throws<ArgumentOutOfRangeException>("offset", () => buffer.Slice(0L, 1L));
+            Assert.Throws<ArgumentOutOfRangeException>("offset", () => buffer.Slice(buffer.Start, 1));
+            Assert.Throws<ArgumentOutOfRangeException>("offset", () => buffer.Slice(buffer.Start, 1L));
+        }
 
-            Assert.Equal(0, c1.GetInteger());
-            Assert.Equal(bufferSegment2, c1.GetObject());
+        [Fact]
+        public void Empty_SliceNegativeLength()
+        {
+            ReadOnlySequence<byte> buffer = ReadOnlySequence<byte>.Empty;
+            Assert.Throws<ArgumentOutOfRangeException>("length", () => buffer.Slice(0, -1));
+            Assert.Throws<ArgumentOutOfRangeException>("length", () => buffer.Slice(0L, -1L));
+            Assert.Throws<ArgumentOutOfRangeException>("length", () => buffer.Slice(buffer.Start, -1));
+            Assert.Throws<ArgumentOutOfRangeException>("length", () => buffer.Slice(buffer.Start, -1L));
+        }
 
-            // Go out of bounds for segment
-            Assert.Throws<ArgumentOutOfRangeException>(() => c1 = buffer.GetPosition(150, buffer.Start));
-            Assert.Throws<ArgumentOutOfRangeException>(() => c1 = buffer.GetPosition(250, buffer.Start));
+        [Fact]
+        public void Empty_Enumerator()
+        {
+            ReadOnlySequence<byte> buffer = ReadOnlySequence<byte>.Empty;
+            ReadOnlySequence<byte>.Enumerator enumerator = buffer.GetEnumerator();
+            {
+                Assert.Equal(default, enumerator.Current); 
+                Assert.True(enumerator.MoveNext());
+                ReadOnlyMemory<byte> memory = enumerator.Current;
+                Assert.True(memory.IsEmpty);
+
+                Assert.False(enumerator.MoveNext());
+            }
+            enumerator = new ReadOnlySequence<byte>.Enumerator(buffer);
+            {
+                Assert.Equal(default, enumerator.Current); 
+                Assert.True(enumerator.MoveNext());
+                ReadOnlyMemory<byte> memory = enumerator.Current;
+                Assert.True(memory.IsEmpty);
+
+                Assert.False(enumerator.MoveNext());
+            }
         }
 
         [Fact]
-        public void Create_WorksWithArray()
+        public void Empty_GetPosition()
         {
-            var buffer = new ReadOnlySequence<byte>(new byte[] { 1, 2, 3, 4, 5 });
-            Assert.Equal(buffer.ToArray(), new byte[] { 1, 2, 3, 4, 5 });
+            ReadOnlySequence<byte> buffer = ReadOnlySequence<byte>.Empty;
+            Assert.Equal(buffer.Start, buffer.GetPosition(0));
+            Assert.Equal(buffer.Start, buffer.GetPosition(0, buffer.Start));
+            Assert.Equal(buffer.Start, buffer.GetPosition(0, buffer.End));
+            //Assert.Equal(buffer.End, buffer.GetPosition(0));
```

Comment:
remove commented out code

Reply:
Here I have question. Should Start equals to End if length equal to 0? Now it is not because we have flags in Start and End.

Reply:
Assert.Equal(buffer.End, buffer.GetPosition(0)) is not equal because buffer.End contains flags. So is it correct to return flags?

Reply:
I think they should be equal, but I will defer it to Pavel. cc @pakrym

Reply:
The question is should Start and End return internal flags or we should remove flags on this properties?

Reply:
> The question is should Start and End return internal flags

Whenever we set _sequenceStart, _sequenceEnd, we adjust the values using the mask:
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/ReadOnlySequence.cs#L86-L87

I think we should remove the mask in those properties (by doing an `and` operation with `IndexBitMask` to clear the mask) so that the values of the Start and End properties make sense to the caller.

Reply:
> I think we should remove the mask in those properties (by doing an and operation with IndexBitMask to clear the mask) so that the values of the Start and End properties make sense to the caller.

Yes my question is about this. I think we should remove flags in this properties but may be we return flags for speed or something else?

Reply:
I can make such changes if it is correct.

Reply:
@pakrym, @davidfowl, can you confirm that it is OK make this change (i.e. remove the flags before returning the value from the Start/End properties)?

Reply:
We don't want customers to interpret SequencePositions at all because we don't have any guarantees about them, including that start == end for empty sequence.

Reply:
@AlexRadch, looking more closely at the use of these properties, we can't (and shouldn't) remove the bit masks. We rely on those flags being set on the SequencePosition that the user passes in to several of the public APIs (that take SequencePosition):
For example: https://github.com/dotnet/corefx/blob/f592e887e2349ed52af6a83070c42adb9d26408c/src/System.Memory/src/System/Buffers/ReadOnlySequence.cs#L309

**Therefore, please remove the commented out code.**

From the documentation for SequencePosition:
`Properties of this type should not be interpreted by anything but the type that created it.`

Reply:
> We rely on those flags being set on the SequencePosition that the user passes in to several of the public APIs (that take SequencePosition)

At first `TryGetBuffer` is not public API, at second `TryGetBuffer` removes this flags before works and get flags from private fields _startPosition and _endPosition.

> `Properties of this type should not be interpreted by anything but the type that created it.`
I completely support this. But I think that it is usefully to support equality comparison between  positions. It is not properties interpretation. It is just equality in test code.

Reply:
> We don't want customers to interpret SequencePositions at all because we don't have any guarantees about them, including that start == end for empty sequence.

SequencePosition supports equality comparison but ROS doesn't. I think it useful to support it.

Reply:
> At first `TryGetBuffer` is not public API, at second `TryGetBuffer` removes this flags before works and get flags from private fields _startPosition and _endPosition.

TryGetBuffer is called from the public API TryGet (where position is passed in by the user).

But I see your point. We don't use the SequencePosition passed in by the user to infer the SequenceType enum, but instead, we call the property. We could just as easily use the private fields instead (_sequenceStart.GetInteger(), _sequenceEnd.GetInteger()), and have the property clear the flags before returning.

https://github.com/dotnet/corefx/blob/f592e887e2349ed52af6a83070c42adb9d26408c/src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs#L26
https://github.com/dotnet/corefx/blob/f592e887e2349ed52af6a83070c42adb9d26408c/src/System.Memory/src/System/Buffers/ReadOnlySequence.cs#L341

@pakrym - thoughts?

cc @GrabYourPitchforks

Reply:
> and have the property clear the flags before returning.
Why? Who would consume the data?

Reply:
> Who would consume the data?

Expecting `start == end for empty sequence` is separate from explicitly getting and interpreting the integer fields, no?

Reply:
> Why? Who would consume the data?

The stupid code with comparison of sequence positions. At second now we clear this flags in each ROS methods. If we removed them, ROS code will be a little faster.